app/vmagent/opentelemetry/request_handler.go

```diff
@@ -22,21 +23,24 @@ var (
 )
 
 // InsertHandler processes opentelemetry metrics.
-func InsertHandler(at *auth.Token, req *http.Request) error {
+func InsertHandler(at *auth.Token, req *http.Request) ([]byte, error) {
+	var respBody []byte
 	extraLabels, err := parserCommon.GetExtraLabels(req)
 	if err != nil {
-		return err
+		return respBody, err
 	}
 	isGzipped := req.Header.Get("Content-Encoding") == "gzip"
 	var processBody func([]byte) ([]byte, error)
 	if req.Header.Get("Content-Type") == "application/json" {
-		if req.Header.Get("X-Amz-Firehouse-Protocol-Version") != "" {
+		if req.Header.Get("X-Amz-Firehose-Protocol-Version") != "" {
```

Comment:
I'd suggest to simplify it a bit:

 create a function at firehose package:

```go
// ResponseWriter writes Ok response for firehose http endpoint request
// https://docs.aws.amazon.com/firehose/latest/dev/httpdeliveryrequestresponse.html#responseformat
func ResponseOkWriter(w http.ResponseWriter, ct time.Time, requestID string) {
	respBody := fmt.Sprintf(`{ "requestId": %q, "timestamp": "%d"}`, requestID, ct.UnixMilli())
	w.Header().Add("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
	w.Write([]byte(respBody))
}
```

 And use it :

```go
var processBody func([]byte) ([]byte, error)
	writeResponse := func(w http.ResponseWriter) {
		w.WriteHeader(http.StatusOK)
	}
	if req.Header.Get("Content-Type") == "application/json" {
		if fhRequestID := req.Header.Get("X-Amz-Firehose-Request-Id"); fhRequestID != "" {
			processBody = firehose.ProcessRequestBody
			writeResponse = func(w http.ResponseWriter) {
				firehose.ResponseOkWriter(w, time.Now(), fhRequestID)
			}
		} else {
			return fmt.Errorf("json encoding isn't supported for opentelemetry format. Use protobuf encoding")
		}
	}
```

Reply:
@f41gh7 please review again